components/raftstore/src/store/async_io/write.rs

```diff
@@ -0,0 +1,640 @@
+            for task in &self.tasks {
+                for t in &task.request_times {
+                    metrics
+                        .write_end
+                        .observe(duration_to_sec(now.saturating_duration_since(*t)))
+                }
+            }
+        }
+    }
+}
+
+struct Worker<EK, ER, T, N>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    T: Transport,
+    N: Notifier,
+{
+    store_id: u64,
+    tag: String,
+    engines: Engines<EK, ER>,
+    receiver: Receiver<WriteMsg<EK, ER>>,
+    notifier: N,
+    trans: T,
+    batch: WriteTaskBatch<EK, ER>,
+    cfg_tracker: Tracker<Config>,
+    raft_write_size_limit: usize,
+    metrics: StoreWriteMetrics,
+    message_metrics: RaftSendMessageMetrics,
+    perf_context: EK::PerfContext,
+}
+
+impl<EK, ER, T, N> Worker<EK, ER, T, N>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    T: Transport,
+    N: Notifier,
+{
+    fn new(
+        store_id: u64,
+        tag: String,
+        engines: Engines<EK, ER>,
+        receiver: Receiver<WriteMsg<EK, ER>>,
+        notifier: N,
+        trans: T,
+        cfg: &Arc<VersionTrack<Config>>,
+    ) -> Self {
+        let batch = WriteTaskBatch::new(
+            engines.kv.write_batch_with_cap(KV_WB_DEFAULT_SIZE),
+            engines.raft.log_batch(RAFT_WB_DEFAULT_SIZE),
+        );
+        let perf_context = engines
+            .kv
+            .get_perf_context(cfg.value().perf_level, PerfContextKind::RaftstoreStore);
+        let cfg_tracker = cfg.clone().tracker(tag.clone());
+        Self {
+            store_id,
+            tag,
+            engines,
+            receiver,
+            notifier,
+            trans,
+            batch,
+            cfg_tracker,
+            raft_write_size_limit: cfg.value().raft_write_size_limit.0 as usize,
+            metrics: StoreWriteMetrics::new(cfg.value().waterfall_metrics),
+            message_metrics: Default::default(),
+            perf_context,
+        }
+    }
+
+    fn run(&mut self) {
+        loop {
+            let loop_begin = Instant::now();
+            let mut handle_begin = loop_begin;
+
+            let mut first_time = true;
+            while self.batch.get_raft_size() < self.raft_write_size_limit {
+                let msg = if first_time {
+                    first_time = false;
+                    match self.receiver.recv() {
+                        Ok(msg) => {
+                            handle_begin = Instant::now();
+                            msg
+                        }
+                        Err(_) => return,
+                    }
+                } else {
+                    match self.receiver.try_recv() {
+                        Ok(msg) => msg,
+                        Err(TryRecvError::Empty) => break,
+                        Err(TryRecvError::Disconnected) => return,
+                    }
+                };
+                match msg {
+                    WriteMsg::Shutdown => return,
```

Comment:
Should we add some graceful shutdown mechanisms? like handle all received tasks before shutdown.

Reply:
Make sense.